### PR TITLE
feat: Add authorization for theme builder by GitHub

### PR DESCRIPTION
**Pull request by neretin-trike**
:baby_chick: Download canary assets:

[colors--canary.281.4023354940.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.281.4023354940.xml)  
[color_metro_ios-swift--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_ios-swift--canary.281.4023354940.swift)  
[color_metro_kotlin--canary.281.4023354940.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_kotlin--canary.281.4023354940.kt)  
[color_metro_react-native--canary.281.4023354940.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_react-native--canary.281.4023354940.ts)  
[color_sbermarket_ios-swift--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.281.4023354940.swift)  
[color_sbermarket_kotlin--canary.281.4023354940.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.281.4023354940.kt)  
[color_sbermarket_react-native--canary.281.4023354940.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.281.4023354940.ts)  
[color_sberprime_ios-swift--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.281.4023354940.swift)  
[color_sberprime_kotlin--canary.281.4023354940.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.281.4023354940.kt)  
[color_sberprime_react-native--canary.281.4023354940.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.281.4023354940.ts)  
[color_selgros_ios-swift--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_ios-swift--canary.281.4023354940.swift)  
[color_selgros_kotlin--canary.281.4023354940.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_kotlin--canary.281.4023354940.kt)  
[color_selgros_react-native--canary.281.4023354940.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_react-native--canary.281.4023354940.ts)  
[color_smbusiness_ios-swift--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_ios-swift--canary.281.4023354940.swift)  
[color_smbusiness_kotlin--canary.281.4023354940.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_kotlin--canary.281.4023354940.kt)  
[color_smbusiness_react-native--canary.281.4023354940.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_react-native--canary.281.4023354940.ts)  
[PlasmaTokensColor--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.281.4023354940.swift)  
[typo_mage_ios-swift--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.281.4023354940.swift)  
[typo_mage_kotlin--canary.281.4023354940.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.281.4023354940.kt)  
[typo_mage_react-native--canary.281.4023354940.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.281.4023354940.ts)  
[typo_plasma_ios-swift--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.281.4023354940.swift)  
[typo_plasma_kotlin--canary.281.4023354940.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.281.4023354940.kt)  
[typo_plasma_react-native--canary.281.4023354940.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.281.4023354940.ts)  
[typo_ruler_ios-swift--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.281.4023354940.swift)  
[typo_ruler_kotlin--canary.281.4023354940.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.281.4023354940.kt)  
[typo_ruler_react-native--canary.281.4023354940.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.281.4023354940.ts)  
[typo_sage_ios-swift--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.281.4023354940.swift)  
[typo_sage_kotlin--canary.281.4023354940.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.281.4023354940.kt)  
[typo_sage_react-native--canary.281.4023354940.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.281.4023354940.ts)  
[typo_sbermarket_ios-swift--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.281.4023354940.swift)  
[typo_sbermarket_kotlin--canary.281.4023354940.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.281.4023354940.kt)  
[typo_sbermarket_react-native--canary.281.4023354940.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.281.4023354940.ts)  
[typo_soulmate_ios-swift--canary.281.4023354940.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.281.4023354940.swift)  
[typo_soulmate_kotlin--canary.281.4023354940.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.281.4023354940.kt)  
[typo_soulmate_react-native--canary.281.4023354940.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.281.4023354940.ts)

### CODE REVIEW

File: website/plasma-theme-builder/src/api/githubAuth.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+export const getAuthToken = async (url: string, pluginClientId: string): Promise<string> => {
+    const token = window.localStorage.getItem('token');
+
+    if (token) {
+        return token;
+    }
+
+    const response = await fetch(`${`${url}/token`}/${pluginClientId}`);
+    const result = await response.json();
```
Comment (Yeti-or) [inline website/plasma-theme-builder/src/api/githubAuth.ts:9]: добавь плиз try catch

File: website/plasma-theme-builder/src/components/AuthRequestModal/AuthRequestModal.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    width: 40rem;
+`;
+
+const StyledButtons = styled.div`
+    display: flex;
+    justify-content: flex-end;
+    gap: 1rem;
+`;
+
+const Description = styled.p``;
```
Comment (Yeti-or) [inline website/plasma-theme-builder/src/components/AuthRequestModal/AuthRequestModal.tsx:15]: тут не нужно взять  Body из типографики ?

File: website/plasma-theme-builder/src/components/Footer/Footer.tsx [hunk 3]
```diff
@@ -50,7 +52,7 @@ export const Footer = () => {
                 </StyledLink>
                 <IconTelegram />
             </Content>
-            <Copyright>©2022 Девайсы</Copyright>
+            <Copyright>©{currentYear} Девайсы</Copyright>
```
Comment (Yeti-or) [inline website/plasma-theme-builder/src/components/Footer/Footer.tsx:55]: =) вот и год прошел

File: website/plasma-theme-builder/src/components/hooks.tsx [hunk 4]
```diff
@@ -68,3 +70,29 @@ export const useNormalizeThemeSections = (data?: ThemeType): ThemeType | undefin
             { ...data },
         );
     }, [data]);
+
+export const useGithubAuth = () => {
+    const {
+        APPLICATION_CLIENT_ID = '',
+        AUTH_SERVER_URL = '',
+        GITHUB_AUTH_API_URL = 'https://github.com/login/oauth/authorize',
```
Comment (Yeti-or) [inline website/plasma-theme-builder/src/components/hooks.tsx:78]: а мы не хотим это на уровень /api/ положить ?

File: website/plasma-theme-builder/src/components/hooks.tsx [hunk 5]
```diff
@@ -68,3 +70,29 @@ export const useNormalizeThemeSections = (data?: ThemeType): ThemeType | undefin
             { ...data },
         );
     }, [data]);
+
+export const useGithubAuth = () => {
+    const {
+        APPLICATION_CLIENT_ID = '',
```
Comment (Yeti-or) [inline website/plasma-theme-builder/src/components/hooks.tsx:76]: а ты их задал уже где-то в env ?

File: website/plasma-theme-builder/src/components/hooks.tsx [hunk 6]
```diff
@@ -68,3 +70,29 @@ export const useNormalizeThemeSections = (data?: ThemeType): ThemeType | undefin
...
+    }, [AUTH_SERVER_URL, pluginClientId]);
+
+    const onGetAuth = useCallback(() => {
+        const redirectUri = `${AUTH_SERVER_URL}/auth/${pluginClientId}`;
+
+        window.location.href = `${GITHUB_AUTH_API_URL}?client_id=${APPLICATION_CLIENT_ID}&redirect_uri=${redirectUri}&response_type=code&scope=public_repo`;
+    }, [APPLICATION_CLIENT_ID, AUTH_SERVER_URL, GITHUB_AUTH_API_URL, pluginClientId]);
+
+    return [token, onGetToken, onGetAuth] as const;
+};
```
Comment (Yeti-or) [inline website/plasma-theme-builder/src/components/hooks.tsx:98]: не хочешь хуки разложить по разным файлам?

**Review by Yeti-or — APPROVED** [review]

File: website/plasma-theme-builder/src/components/Token/Token.tsx [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    align-items: center;
+    justify-content: space-between;
+
+    width: 60rem;
+
+    margin: 1rem 0;
+    margin-left: 2rem;
+
+    ${({ enabled }) =>
+        enabled === false &&
```
Comment (Yakutoc) [inline website/plasma-theme-builder/src/components/Token/Token.tsx:34]: @neretin-trike Так сделано потому что **enabled** опциональный проп?

File: website/plasma-theme-builder/src/components/hooks.tsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    const onGetToken = useCallback(async () => {
+        const token = await getAuthToken(AUTH_SERVER_URL, pluginClientId);
+
+        setToken(token);
+    }, [AUTH_SERVER_URL, pluginClientId]);
+
+    const onGetAuth = useCallback(() => {
+        const redirectUri = `${AUTH_SERVER_URL}/auth/${pluginClientId}`;
+
+        window.location.href = `${GITHUB_AUTH_API_URL}?client_id=${APPLICATION_CLIENT_ID}&redirect_uri=${redirectUri}&response_type=code&scope=public_repo`;
```
Comment (Yakutoc) [inline website/plasma-theme-builder/src/components/hooks.tsx:94]: [Вкусовщина]

@neretin-trike А что если URL делать через `new URL(url)` в сочетании с `URLSearchParams`

**Review by Yakutoc — APPROVED** [review]

File: website/plasma-theme-builder/src/components/Token/Token.tsx (hunk 7 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/components/Token/Token.tsx:34]: ага, но при этом если он undefined или true, то данный стиль примениться не должен

File: website/plasma-theme-builder/src/components/AuthRequestModal/AuthRequestModal.tsx (hunk 2 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/components/AuthRequestModal/AuthRequestModal.tsx:15]: да, можно и так 👌

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-281/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-281/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-281/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-281/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-281/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-281/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-281/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-281/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-281/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-281/

File: website/plasma-theme-builder/src/components/hooks.tsx (hunk 5 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/components/hooks.tsx]: вообще да, добавил. Но либо не туда, либо чото не доделал, тут нужна твоя консультация

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-281/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-281/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-281/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-281/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-281/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-281/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-281/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-281/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-281/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-281/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-281/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-281/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-281/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-281/temple-storybook/

**Pull request merged**